shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2613,150 +2628,166 @@ impl KafkaSinkCluster {
+                                &rack.as_ref(),
+                            );
+                            let broker_ids = collect_broker_ids(shotover_nodes_by_rack);
+                            shotover_isr_nodes.extend(broker_ids);
+                        } else {
+                            // If the cluster detects the broker is down it will not be included in the metadata list of brokers.
+                            // In that case we wont find a broker and we should just skip this broker.
+                        }
+                    }
+                    partition.isr_nodes = shotover_isr_nodes.into_iter().collect();
 
-                // TODO: handle this properly, for now its better to just clear it than do nothing
-                partition.offline_replicas.clear();
+                    // TODO: handle this properly, for now its better to just clear it than do nothing
+                    partition.offline_replicas.clear();
+                }
             }
-        }
 
-        if let Some(controller_node) = self
-            .nodes
-            .iter()
-            .find(|node| node.broker_id == metadata.controller_id)
-        {
-            // If broker has no rack - use the first shotover node
-            // If broker has rack - use the first shotover node with the same rack
-            // This is deterministic because the list of shotover nodes is sorted.
-            if let Some(shotover_node) = self.shotover_nodes.iter().find(|shotover_node| {
-                controller_node
-                    .rack
-                    .as_ref()
-                    .map(|rack| rack == &shotover_node.rack)
-                    .unwrap_or(true)
-            }) {
-                metadata.controller_id = shotover_node.broker_id;
-            } else {
-                tracing::warn!(
-                    "No shotover node configured to handle kafka rack {:?}",
-                    controller_node.rack
+            if let Some(controller_node_rack) = self
+                .nodes
+                .iter()
+                .find(|node| node.broker_id == metadata.controller_id)
+                .map(|x| x.rack.clone())
+            {
+                // If broker has no rack - use the first UP shotover node.
+                // If broker has rack - use the first UP shotover node with the same rack if available,
+                // and fall back to use the first UP shotover node out of the rack otherwise.
+                // This is deterministic because the list of UP shotover nodes is sorted and partitioning does not change the order.
+                let shotover_nodes_by_rack = partition_shotover_nodes_by_rack(
+                    &up_shotover_nodes,
+                    &controller_node_rack.as_ref(),
                 );
+                let shotover_node = shotover_nodes_by_rack
+                    .nodes_in_rack
+                    .first()
+                    .or_else(|| shotover_nodes_by_rack.nodes_out_of_rack.first());
```

Comment:
lets use an `.expect("There will always be at least one up node")` to avoid needing to handle an error that wont happen.

Reply:
added